python/rmm/_lib/memory_resource.pxd

```diff
@@ -5,100 +5,128 @@ from libcpp.vector cimport vector
 from libcpp.string cimport string
 from libcpp.memory cimport shared_ptr
 
-
-cdef extern from "memory_resource_wrappers.hpp" nogil:
-    cdef cppclass device_memory_resource_wrapper:
-        shared_ptr[device_memory_resource_wrapper] get_mr() except +
-
-    cdef cppclass default_memory_resource_wrapper(
-        device_memory_resource_wrapper
-    ):
-        default_memory_resource_wrapper(int device) except +
-
-    cdef cppclass cuda_memory_resource_wrapper(device_memory_resource_wrapper):
-        cuda_memory_resource_wrapper() except +
-
-    cdef cppclass managed_memory_resource_wrapper(
-        device_memory_resource_wrapper
-    ):
-        managed_memory_resource_wrapper() except +
-
-    cdef cppclass pool_memory_resource_wrapper(device_memory_resource_wrapper):
-        pool_memory_resource_wrapper(
-            shared_ptr[device_memory_resource_wrapper] upstream_mr,
-            size_t initial_pool_size,
-            size_t maximum_pool_size
-        ) except +
-
-    cdef cppclass fixed_size_memory_resource_wrapper(
-        device_memory_resource_wrapper
-    ):
-        fixed_size_memory_resource_wrapper(
-            shared_ptr[device_memory_resource_wrapper] upstream_mr,
+cdef extern from "thrust/optional.h" namespace "thrust" nogil:
+
+    struct nullopt_t:
+        pass
+
+    cdef nullopt_t nullopt
+
+    cdef cppclass optional[T]:
+        optional()
+        optional(T v)
+
+    cdef optional[T] make_optional[T](T v)
```

Comment:
We're now leaking Thrust as a dependency if anyone wants to use RMM's Cython, any other options?

Reply:
We added optional parameters to `pool_memory_resource` on the C++ side a while ago, using `Thrust::optional`. I didn't thnk about the fact that it would leak Thrust. RMM is a header-only library so it leaks a Thrust dependency no matter what (on the C++ side, at least).

Reply:
On the C++ side we could choose to install our own copy of Thrust alongside the RMM headers and use them within RMM (I'm not advocating for this). On the Cython, anyone who uses RMM's Cython will then need to be able to find Thrust headers in addition to the RMM headers as part of their build process and its generally just messier from the Python side.

Reply:
I agree. We may want to do some Cython shenanigans to hide it, if possible.

Reply:
I have moved most of the extern declarations into the implementation file to avoid leaking the dependencies. This allows the build to work without having any unnecessary dependencies in the .pxd file.

If this is different than what you were expecting. Let me know.

Reply:
This looks like a good tradeoff to me. We can revisit if/when we move to using C++17 in the future and can use `std::optional` instead of `thrust::optional`.